bip-0352.mediawiki

```diff
@@ -0,0 +1,476 @@
+
+Otherwise, skip the transaction.
+
+=== Address encoding ===
+
+A silent payment address is constructed in the following manner:
+
+* Let ''B<sub>scan</sub>, b<sub>scan</sub> = Receiver's scan public key and corresponding private key''
+* Let ''B<sub>spend</sub>, b<sub>spend</sub> = Receiver's spend public key and corresponding private key''
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G'', where ''m'' an optional integer tweak for labeling
+** In the case of ''m'' = 0, no label is applied and ''B<sub>m</sub> = B<sub>spend</sub>''
+* The final address is a [https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki Bech32m] encoding of:
+** The human-readable part "sp" for mainnet, "tsp" for testnets (e.g.  signet, testnet)
+** The data-part values:
+*** The character "q", to represent a silent payment address of version 0
+*** The 66 byte concatenation of the receiver's public keys, ''ser<sub>P</sub>(B<sub>scan</sub>) || ser<sub>P</sub>(B<sub>m</sub>)''
+
+Note: [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki BIP173] imposes a 90 character limit for Bech32 strings, whereas a silent payment address requires ''at least'' 117 characters<ref name="why_117_chars"> ''' Why do silent payment addresses need at least 117 characters?''' A silent payment address is a bech32m encoding comprised of the following parts:
+
+
+* HRP [2-3 characters]
+* separator [1 character]
+* version [1-2 characters]
+* payload, 66 bytes concatenated pubkeys [ceil(66*8/5) = 106 characters]
+* checksum [6 characters]
+
+
+For a silent payments v0 address, this results in a 117 character address when using a 3 character HRP. Future versions of silent payment addresses may add to the payload, which is why a 1023 character limit is suggested.</ref>. Additionally, since higher versions may add to the data field, it is recommended implementations use a limit of 1023 characters (see [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#checksum-design BIP173: Checksum design] for more details).
+
+=== Outpoints hash ===
+
+The sender and receiver MUST calculate an outpoints hash for the transaction in the following manner:
+
+* Collect each ''outpoint'' used as an input to the transaction
+* Let ''outpoints = outpoint<sub>0</sub> || ... || outpoint<sub>n</sub>'', sorted lexicographically by txid and vout, ascending order<ref name="why_sort_outpoints">'''Why are outpoints sorted before hashing?''' This way the silent payment output does not need to be recalculated if the wallet changes the order of inputs, e.g. at signing time or during an RBF bump. Note that the silent payment output does need to be recalculated if inputs are added or removed.</ref>
+* Let ''outpoints_hash = sha256(outpoints)''
```

Comment:
This is problematic for hardware wallets. Hardware wallets are constrained in memory and may not be able to load all outpoints and sort them. As a result, they would have to restrict transactions to already have sorted inputs at signing time, which is not desirable and would lead to compatibility issues.

Maybe instead of using the hash of the sorted outpoints, we could use something like:

`sha256(outpoint_0) + sha256(outpoint_1) + ... + sha256(outpoint_n)`, which does not depend on the ordering of the outpoints. 

It would be great if a cryptographer could validate if the above or something similar works and if it is safe.

Reply:
Thanks for the suggestion @benma ! I'm leaning toward your suggestion (assuming there are no issues cryptographically). Not needing a sort feels less brittle than needing to worry about having the outpoints in the right order before hashing. The only open question I have is how much would this affect the scanning time for a full-node. I'll make a benchmark and get back to you on that.

Reply:
> it will in general not be preimage resistant

See [Wagner's algorithm](https://www.enseignement.polytechnique.fr/informatique/profs/Francois.Morain/Master1/Crypto/projects/Wagner02.pdf) for the generalized birthday problem.

Reply:
Thanks @sipa , @jonasnick ! For context, we hash the outpoints used in the transaction to come up with a unique scalar to multiply the ECDH shared secret with. The shared secret is then hashed again before creating the final output. The purpose of this scalar is to prevent generating the same `scriptPubKey` for the receiver when spending UTXOs from the same sender scriptPubKeys (e.g. a large exchange spending from a scriptPubKey that has been used multiple times).

What we need is for this scalar to be different for each set of UTXOs that come from a reused scriptPubKey, being sent to the same receiver. If I'm understanding your points correctly, the concern here with hashing each outpoint and summing is that we might end up with the same scalar for two different sets of UTXOs?

Reply:
Thanks @sipa and @jonasnick. 

Correct me if I am wrong @josibake @RubenSomsen, but the point of the hash of the inputs was not preimage resistance, but just to make sure the output pubkey changes when the inputs change to avoid pubkey reuse.

In that case, summing instead of hashing could still work.

Reply:
that's correct @benma regarding the purpose and the fact that the preimage is known. I think the more important question is whether or not this approach is collision-resistant: if hashing and summing a large number of outpoints, what is this risk we end up with the same hash as a different set of outpoints when spending from the same `scriptPubKey`?

Reply:
Thanks - good to know that in the honest setting, it wold work.

> Deliberately constructing a set of say 32 (valid) outpoints such that the sum of their hashes matches a fixed value is completely feasible (I'd guess a few hours of CPU time). Collisions are even easier.

@sipa for clarity: the sender would generate the UTXOs first by locally brute forcing transactions creating such outputs until they result in some fixed value, right?

About the adversarial setting: failure in this case means the outpoints are chosen so that the output key has a collision (output script or address reuse). Since the ECDH shared secret is involved, only the sender can invoke this failure. 

- The sender can do this regardless of the hashing, as they can just send to the same output script using any kind of transaction. The recipient might not find the payment unless they also scan for other transactions paying to a silent payment output. @josibake is a recipient supposed to do that? 
- Does a sender have any motivation to perform this kind of work just to be able to reuse an address (output script) of the recipient? The recipient can detect that this happened easily enough.

Anything I missed about the adversarial case?

Reply:
Regarding the adversarial case, perhaps this is an issue in a collaborative spend (coinjoin): the non-sending participants could craft malicious outpoints such that they force the sender to re-create a scriptPubKey from a previous send. But I don't think this is feasible given that they would need to know ahead of time the recipient address (which we are hiding with [blind DH](https://gist.github.com/RubenSomsen/be7a4760dd4596d06963d67baf140406)) and which UTXOs the honest participant (sender) intends to spend.

Also, as @benma mentioned, we can't stop malicious address reuse since the sender can send to a previously created scriptPubKey at any time. The outpoint hash is only to prevent _honest_ participants from accidentally recreating a scriptPubkey when sending to the same receiver and spending from a scriptPubKey that has been reused.

> The recipient might not find the payment unless they also scan for other transactions paying to a silent payment output. @josibake is a recipient supposed to do that?

This is more of an implementation detail for wallets: does your wallet keep checking for scriptPubKeys it has already learned about. I imagine most wallets already do this, and I would certainly recommend a silent payments wallet also check previously found SP outputs when scanning a new tx/block.

Reply:
After thinking this over some more, our best option is to remove the sorting step. This means that the sender and receiver hash the outpoints in the order they appear in the final transaction. This was our original proposal and the sorting step was added to ensure the silent payments output was not dependent on the transaction structure, as a "belt and suspenders" approach.

But the order of the inputs is committed to when signing the transaction, and in the hardware wallet context, it is during signing that the silent payment output would be created anyway, so I think this is fine. For software wallets, we just need to make it explicit in the BIP that you must not change the order of the inputs *after* generating the silent payments output, e.g. generate the silent payment output when signing the transaction.

The benefits of removing the sort are:

1. Easier for hardware wallets to implement
2. Small improvement in scanning time

Curious to here your thoughts @RubenSomsen , @benma ?

Reply:
Doesn't the hardware wallet need to load the entire transaction in order to sign it anyway? Is the small overhead for sorting the outpoints really a problem?

Reply:
@benma I see, then wouldn't it be possible for the connected software to stream the sorted outpoints as well? Verifying that they're sorted correctly would only require a constant amount of memory.

Reply:
I agree that a simple and clean solution is just to drop the sorting. If without a concrete concern w.r.t. hw wallets, I feel that the sorting adds too much complexity given that it's not clear what it actually achieves.

Here are some more thoughts on this:  I agree with sipa and nickler that summing hashes makes them prone to collisions due to Wagner's attack (similar attacks such as BLLOR). But collision resistance is a goal here because this is what will "prevent generating the same scriptPubKey for the receiver when spending UTXOs from the same sender scriptPubKeys". Or at least, this ensures that when there are transactions with colliding scriptPubKeys for the receiver, only one of them will be valid because they all spend the same UTXOs. 

This hints at another possible solution: Hash only the first outpoint. This ensures that any transactions with colliding scriptPubKeys for the receiver will be double-spends. A transaction doesn't need to double-spend all the inputs to be invalid, one input suffices... ^^ This is even more efficient when it comes to hashing, perhaps at the cost that it looks bit inelegant.

Reply:
If we're hashing only one outpoint (I don't see any immediate problem with that, thanks for the suggestion @real-or-random), we might as well use the lowest (lexicographically) outpoint and retain the benefits of sorting (which are outlined in the footnote).

Reply:
Hi, I'm catching up on the discussion here, I just would like to sum up what I understood of the issue because it's not entirely clear for me. 

If I understand correctly the problem with the way we are using a hash of all outpoints here, is that it is theoretically relatively easy to find 2 sets of different outpoints that will produce the same hash. If we suppose that we also have the same set of keys in the scriptpubkeys those outpoints point to, a malicious payer would be able to make 2 different transactions (no double spend) that pays the same silent payment destination keys, forcing address reuse to the receiver, correct?

I would just confirm that I got this right before I elaborate further.
(Anyway I like the last proposal from @real-or-random it seems to me that it fulfills the purpose of the outpoints hash in a simpler way, but I need to think more about it)

Reply:
> Out of curiosity, why did you reject the alternative of summing the hashes?

I ran a rudimentary benchmark and the hashing and summing technique was quite a bit slower compared to hashing a concatenation of the bytes. Over the course of the conversation, I've also come to agree that hashing and summing or sorting then hashing seem unjustifiably complex, given that it is only to guard against a theoretical implementation bug.

> we might as well use the lowest (lexicographically) outpoint and retain the benefits of sorting (which are outlined in the footnote).

This strikes a nice balance: simple while still retaining the benefits of sorting. @benma curious to hear your thoughts on this. Seems you could stream the lexicographically lowest input first, then stream all the inputs (so you would be streaming N+1 inputs), or could keep track of which input is the lowest until all the inputs have been streamed before proceeding with the silent payments protocol?

Reply:
@Sosthene00 

> If I understand correctly the problem with the way we are using a hash of all outpoints here

If by "here" you mean the newly proposed hash each outpoint and then sum the hashes, then yes: the concern is that it is easier to find collisions and also requires the scanner to do more work. Regarding collisions, this is only an issue in terms of honest collisions, as a malicious actor can just force address reuse by sending to a previously generated scriptPubKey.

The issue with the old solution (sort the outpoints, concatenate, then hash) is that it is generally complex and specifically difficult for memory-constrained devices such as hardware wallets.

Taking all of that into account, I think hashing the lexicographically lowest outpoint is likely the best solution.

Reply:
@josibake using the hash of the lexicographically smallest input would work for hardware wallets, I don't see any problems with this approach.

Reply:
> If by "here" you mean the newly proposed hash each outpoint and then sum the hashes, then yes: the concern is that it is easier to find collisions and also requires the scanner to do more work. Regarding collisions, this is only an issue in terms of honest collisions, as a malicious actor can just force address reuse by sending to a previously generated scriptPubKey.

Sorry I wasn't very clear about the scheme I was talking about, thanks for correcting me. It seemed to me that it was quite a reach for a malicious actor indeed, as it would be quite hard (impossible?) to pull in practice even for say coinjoins, and would achieve little more than simply reusing an existing address from a previous transaction, but I wanted to make sure I didn't miss something. 

> Taking all of that into account, I think hashing the lexicographically lowest outpoint is likely the best solution.

But if we all agree this is better, my question becomes irrelevant anyway.

Reply:
@RubenSomsen looks like rough consensus on using hashing the lowest outpoint (lexicographic). AFAICT, this doesn't cause any issues for collaborative use cases as we were requiring all of the outpoints before, and now we are just picking one from all of the outpoints.